### PR TITLE
[BB2-896] Responsive Design Sample App Python React

### PR DESCRIPTION
**Pull request by JFU-GIT**
**JIRA Ticket:**
[BB2-896](https://jira.cms.gov/browse/BB2-896)

**User Story or Bug Summary:**

As a 3rd party app, I should be able to understand what to do when I receive certain error messages from BB2 so that I can properly respond to my own users

AC:

Draft documentation for review for:

- What to do when receiving a 403 error
- What to do when receiving a 404 error
- What to do when receiving a 500 error
- how to implement exponential backoff

In the Node.JS and Python sample apps add in responsive design to show:

- What to do when receiving a 403 error
- What to do when receiving a 404 error
- What to do when receiving a 500 error
- how to implement exponential backoff

This PR is for applying similar error handling and retry logic to BB2 Sample App Python React.

### What Should Reviewers Watch For?

* Are error responses from API calls handled as instructed by the document?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is the retry work as expected?

If you're reviewing this PR, please check these things, in particular:

* Review the document on API error responses
* Verify the retry logic works (exponential back off)

To facilitate the verification of retry failed requests, an instrumented BB2 server is deployed to sandbox canary:
put below line into your host file on you local machine:

10.246.44.54	sandbox.bluebutton.cms.gov

then verify retry on EOB request by following following steps:

1. create an app on sandbox (or using an existing one)
2. make sure the url is among the redirect urls: http://server:3001/api/bluebutton/callback/
3. copy the client_id, client_secret to the sample app server config.ts
4. start sample app in docker: go to local repo base and run : docker-compose up

Wait until the server and client components are up and running.
note, a waffle switch "simulate_500_error" is provisioned on sandbox which when "active", will produce a 500 error on fhir request, with error message: "Service temporarily unavailable, try again later, network connection refused".

From client webUI, click button "Authorize" to complete the beneficiary medicare.gov login, and grant access to the app, the corresponding EOB search will hit the simulated 500 network connection error, and the failed EOB request met the criteria for retry.

below is a sample server stdout showing the retry process, the 3rd attempt succeeded (by setting switch "simulate_500_error" to false soon after 2nd retry ...)

server_1  | Response status code: 500
server_1  | Response text: {"detail":"Service temporarily unavailable, try again later, network connection r
efused."}
server_1  | retrying started ...
server_1  | retry attempts: 1
server_1  | retry attempts: 2
server_1  | retry attempts: 3
server_1  | retry successful:
server_1  | Response status_code: 200
server_1  | Response stext: {"resourceType":"Bundle","id":"4b33d4ff-08da-4daa-8a54-ae033a395cb8","meta":{"la
stUpdated":"2021-09-23T13:36:42.530-04:00"},"type":"searchset","total":40,"link":[{"relation":"first","url":
.... FHIR bundle truncated for briefness.

### CODE REVIEW

File: ErrorResponses.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+| 500<br>SERVER ERROR | /v[12]/fhir/.* | The root cause of the 500 error, some times, is indicated by the error message, <br>the app can choose to retry the failed request depend on the nature of the root cause, <br>examples that might be retriable are those related to network down (temporarily):<br>Example:<br>ConnectionError at /mymedicare/login<br>HTTPSConnectionPool(host='test.accounts.cms.gov', port=443): <br>Max retries exceeded with url: /health (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f46599dafd0>: <br>Failed to establish a new connection: [Errno -2] Name or service not known')) | Heuristic on Retry | App can choose to retry on some of the 500 errors as shown by the example, this is a heuristic approach. |
+
+
+### Retry
```
Comment (nbragdon) [inline ErrorResponses.md:36]: I might move this section to the top. That way if they don't want to read through the whole table, at least they got the important takeaway.

File: server/src/utils/bb2Util.py [hunk 2]
```diff
@@ -64,6 +67,7 @@ def getBenefitData(settings, configsSettings, query, loggedInUser):
     }
     BB2_BENEFIT_URL = configsSettings.get('bb2BaseUrl') + '/' + settings.version + '/fhir/ExplanationOfBenefit/'
     myHeader = {'Authorization' : 'Bearer '+loggedInUser.get('authToken').get('access_token')}
-    beneResponse = requests.get(url=BB2_BENEFIT_URL,params=PARAMS,headers=myHeader)    
+    beneResponse = requests.get(url=BB2_BENEFIT_URL,params=PARAMS,headers=myHeader)
```
Comment (nbragdon) [inline server/src/utils/bb2Util.py:70]: Is the retry logic baked into the requests library? If so, do we need to remove the retry from the authorization endpoints? So far I'm basically just seeing added logging (which I'm not actually sure we want here)

Comment (JFU-GIT) [reply server/src/utils/bb2Util.py:70]: haven't put in the retry code yet, working on it.

Comment (JFU-GIT) [reply server/src/utils/bb2Util.py]: error response reporting + retry committed.

File: server/src/app/views.py [hunk 3]
```diff
@@ -24,55 +23,50 @@
...
-    settings = Settings(myEnv,myVersion,PKCE)
-
-    configSettings = getConfigSettings(myEnv)
-    authUrl = generateAuthorizeUrl(settings, configSettings)
-    return authUrl
+    return generate_authorize_url(Settings(my_env, my_version, PKCE), get_config_settings(my_env))
 
 @app.route('/api/authorize/currentAuthToken',methods=['GET'])
-def getCurrentAuthToken():
+def get_current_auth_token():
```
Comment (nbragdon) [inline server/src/app/views.py:46]: Nice job changing the style to match the language. This also makes me think of @dtisza1 suggestion to add style enforcing (like black, etc.) and maybe we can get that in a future release. Appreciate you making these updates here!

**Comment by JFU-GIT** [comment]
Closed as repo stale PR cleanup

**Pull request closed without merging**